integration/request-test.ts

```diff
@@ -0,0 +1,162 @@
+  await app.goto("/");
+  await app.clickElement("#set-cookie");
+
+  let loaderData = JSON.parse(await page.locator("#loader-data").innerHTML());
+  expect(loaderData.method).toEqual("GET");
+  expect(loaderData.url).toMatch(/^http:\/\/localhost:\d+\/$/);
+  expect(loaderData.headers.cookie).toEqual(undefined);
+  expect(loaderData.text).toEqual("");
+
+  await app.clickElement("#submit-get-csr");
+
+  loaderData = JSON.parse(await page.locator("#loader-data").innerHTML());
+  expect(loaderData.method).toEqual("GET");
+  expect(loaderData.url).toMatch(/^http:\/\/localhost:\d+\/\?type=csr$/);
+  expect(loaderData.headers.cookie).toEqual("cookie=nomnom");
+  expect(loaderData.text).toEqual("");
+});
+
+test("action + loader requests SSR POST requests", async ({ page }) => {
+  let app = new PlaywrightFixture(appFixture, page);
+  await app.goto("/");
+  await app.clickElement("#set-cookie");
+
+  let loaderData = JSON.parse(await page.locator("#loader-data").innerHTML());
+  expect(loaderData.method).toEqual("GET");
+  expect(loaderData.url).toMatch(/^http:\/\/localhost:\d+\/$/);
+  expect(loaderData.headers.cookie).toEqual(undefined);
+  expect(loaderData.text).toEqual("");
+
+  await app.clickElement("#submit-post-ssr");
+
+  let actionData = JSON.parse(await page.locator("#action-data").innerHTML());
+  expect(actionData.method).toEqual("POST");
+  expect(actionData.url).toMatch(/^http:\/\/localhost:\d+\/$/);
+  expect(actionData.headers.cookie).toEqual("cookie=nomnom");
+  expect(actionData.text).toEqual("type=ssr");
+
+  loaderData = JSON.parse(await page.locator("#loader-data").innerHTML());
+  expect(loaderData.method).toEqual("POST");
+  expect(loaderData.url).toMatch(/^http:\/\/localhost:\d+\/$/);
+  expect(loaderData.headers.cookie).toEqual("cookie=nomnom");
+  expect(loaderData.text).toEqual("");
+});
+
+test("action + loader requests on CSR POST requests", async ({ page }) => {
+  let app = new PlaywrightFixture(appFixture, page);
+  await app.goto("/");
+  await app.clickElement("#set-cookie");
+
+  let loaderData = JSON.parse(await page.locator("#loader-data").innerHTML());
+  expect(loaderData.method).toEqual("GET");
+  expect(loaderData.url).toMatch(/^http:\/\/localhost:\d+\/$/);
+  expect(loaderData.headers.cookie).toEqual(undefined);
+  expect(loaderData.text).toEqual("");
+
+  await app.clickElement("#submit-post-csr");
+
+  let actionData = JSON.parse(await page.locator("#action-data").innerHTML());
+  expect(actionData.method).toEqual("POST");
+  expect(actionData.url).toMatch(/^http:\/\/localhost:\d+\/$/);
+  expect(actionData.headers.cookie).toEqual("cookie=nomnom");
+  expect(actionData.text).toEqual("type=csr");
+
+  loaderData = JSON.parse(await page.locator("#loader-data").innerHTML());
+  expect(loaderData.method).toEqual("GET");
+  expect(loaderData.url).toMatch(/^http:\/\/localhost:\d+\/$/);
+  expect(loaderData.headers.cookie).toEqual("cookie=nomnom");
+  expect(loaderData.text).toEqual("");
```

Comment:
During SPA submissions, the loader is called through separate revalidation `GET` requests, so the assertions here differ on the method.

I think this all makes sense given how Remix is executing loaders differently in MPA/SPA calls, even if it introduces a slight different between them from a DX standpoint.  Users shouldn't be branching on `request.method` in loaders anyway?

Reply:
This logic is implemented in the router here: https://github.com/remix-run/react-router/pull/9721